src/accelerate/data_loader.py

```diff
@@ -388,9 +389,65 @@ def end(self):
         self.gradient_state._remove_dataloader(self)
 
 
-class DataLoaderShard(DataLoader, DataLoaderStateMixin):
+class DataLoaderAdapter:
     """
-    Subclass of a PyTorch `DataLoader` that will deal with device placement and current distributed setup.
+    A class which wraps around a PyTorch `DataLoader` (or variants of it) to be used with the `Accelerator`. For
+    compatability reasons, this class inherits from the class it wraps around, so it can be used as a drop-in.
+    """
+
+    def __init__(self, dataset, use_stateful_dataloader=False, batch_sampler=None, **kwargs):
+        self.use_stateful_dataloader = use_stateful_dataloader
+        if is_torchdata_stateful_dataloader_available():
+            from torchdata.stateful_dataloader import StatefulDataLoader
+
+        if use_stateful_dataloader and not is_torchdata_stateful_dataloader_available():
+            raise ImportError("StatefulDataLoader is not available. Please install torchdata to use it.")
```

Comment:
Indicate min version of torchdata.

Reply:
Done, I will once again warn that this depends on the nightly build of torchdata.

Reply:
Just checked, it turns out torchdata officially supports this feature from version 0.8.0 onwards: https://github.com/pytorch/data/releases/tag/v0.8.0

Updated docs for this, can probably officially support this without some nightly hacks now